### PR TITLE
ADR-082: Data Companion API

### PR DESCRIPTION
**Pull request by thanethomson**
:book: [Rendered](https://github.com/tendermint/tendermint/blob/thane/adr-082-data-companion-api/docs/architecture/adr-082-data-companion-api.md)

This ADR aims to present an alternative to [ADR-075](https://github.com/tendermint/tendermint/blob/main/docs/architecture/adr-075-rpc-subscription.md). I'm also still trying to understand the degree to which the proposed solution here is feasible (and if not, what would need to change?), but there's some time pressure to make some key architectural decisions given our upcoming Q4 planning and various users' desires to have ADR-075 implemented so I haven't had as much opportunity to do due diligence as I'd hoped. Feedback here would be greatly appreciated.

If it'll work, I imagine this is by no means the only possible solution to the problem outlined in the ADR and I'm totally open to other options here. If we end up going an alternative route, I'd recommend we still capture this in the `docs/architecture` folder and mark it as rejected, so at least the reasoning's captured for posterity, and in case some of it becomes relevant in future.

cc @adizere @romac @ancazamfir @JayT106 

---

#### PR checklist

- [x] Tests written/updated, or no tests needed
- [x] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [x] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

**Comment by adizere** [comment]
Is it an appropriate mental model to consider an IBC relayer as implementing `DataCompanionService` ? Or is it more appropriate to consider some middleware service as implementing `DataCompanionService` which would be _sitting between_ Tendermint and the IBC relayer?

**Comment by thanethomson** [comment]
> Is it an appropriate mental model to consider an IBC relayer as implementing `DataCompanionService` ? Or is it more appropriate to consider some middleware service as implementing `DataCompanionService` which would be _sitting between_ Tendermint and the IBC relayer?

It depends on the overall operational architecture of your setup. The main idea is to attach this `DataCompanionService` to a full node, and it would get all blocks and execution results (including events), without any filtering. The companion would handle any filtering, storage and republishing as necessary.

If your full node needs a scalable, queryable RPC _and_ to facilitate relaying, then you'd need some kind of intermediate node to facilitate both.

If your full node just needs to facilitate relaying, you could just attach the relayer to this endpoint and discard whatever data you don't need.

**Comment by JayT106** [comment]
> The companion would handle any filtering, storage and republishing as necessary.

Do you mean the companion server or the Tendermint side? Not sure the data scales will be for every `CommitBlockRequest`. Will it be better if we can filter the events from the Tendermint side? i.e. the current indexer uses `index-events` in the SDK `app.toml` to filter the events that don't need to be indexed.

Who will implement and maintain the `DataCompanionService` middleware if we introduce this ADR? Does it move the responsibility to the cosmos app project? Feels like those [APIs](https://github.com/tendermint/tendermint/blob/thane/adr-082-data-companion-api/docs/architecture/adr-082-data-companion-api.md#implications) might still be needed within DataCompanionService.

The current Tendermint support offline reindex CLI, will it be needed as well?

File: docs/architecture/adr-082-data-companion-api.md [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+  string codespace = 8;
+}
+```
+
+### Request Buffering
+
+In order to ensure reliable delivery, while also catering for intermittent
+faults, we should allow for configurable buffering of a reasonable, yet small,
+number of requests to the companion service. If this buffer fills up, we should
+panic.
```
Comment (williambanfield) [inline docs/architecture/adr-082-data-companion-api.md:224]: Hmm, I would conceptualize this as backpressure. It doesn't cause unbounded memory growth and it doesn't hang the running goroutines responsible for consensus. Instead, it halts them completely along with all other routines in the program. The only distinction I see is really whether or not the process as a whole is running. In both versions of backpressure (current vs proposed here), the consensus goroutines are stopped once the buffer fills up. In the current version, they _may_ resume on their own without operator intervention. 

By performing backpressure in this way, we also lose the ability to perform the other functions of the node once the buffer has filled because the node will have crashed. In the current scheme, backpressure stalls consensus, but the other facets of the node: p2p, statesync, blocksync etc. are still available.

**Comment by thanethomson** [comment]
> > The companion would handle any filtering, storage and republishing as necessary.
>
> Do you mean the companion server or the Tendermint side?

The companion server. The companion server would get _all_ data and then it would need to decide what to keep and what to discard. One of the options outlined in the ADR is to be able to specify a single query in the [configuration]https://github.com/tendermint/tendermint/blob/thane/adr-082-data-companion-api/docs/architecture/adr-082-data-companion-api.md#configuration), but that seems like an optimization to me and I wouldn't want to prematurely optimize here.

>Not sure the data scales will be for every CommitBlockRequest. Will it be better if we can filter the events from the Tendermint side?

I'm not sure I understand @JayT106. Could you please clarify here?

File: docs/architecture/adr-082-data-companion-api.md [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+2. All block data (i.e. `Block` data structures that have been committed by the
+   consensus engine).
+
+## Alternative Approaches
+
+One clear alternative to this would be the approach outlined in [ADR
+075][adr-075]. This approach still unfortunately leaves Tendermint responsible
+for maintaining a query interface and event indexing functionality, increasing
+the long-term maintenance burden of, and the possibility of feature sprawl in,
+that subsystem.
```
[the remark this answers is not in the document]
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:63]: We could do that, but it doesn't solve all the problems this ADR aims to solve.

1. The solution in ADR-075 only keeps track of a sliding window of events (and does so in memory), and so there's still the possibility that subscribers who need guaranteed event delivery (e.g. relayers) may end up losing events in the case of intermittent network failures.
2. We would need to establish a way to impose a single-consumer limit in order to be able to automatically prune stored events and block results. The assumption in this ADR is that, because we have a single reliable companion, if we've sent them the data we won't need to re-send it.
3. We would still be offering the data via JSON as opposed to Protobuf. This may have performance and bandwidth implications with this "firehose" design.

If we just modified ADR-075 to address these issues, we would probably converge on the same approach outlined in this ADR except we'd be sending the data via JSON over HTTP instead of Protobuf over HTTP/2.

File: docs/architecture/adr-082-data-companion-api.md (hunk 1 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:224]: Think of this model as providing a dedicated, reliable "store" for the data we're starting off sending with this ADR. Failing to send this data effectively would be the equivalent of failing to save ABCI responses [here](https://github.com/tendermint/tendermint/blob/0c96f0b434c1d9f23883d1f2291135830a7a5290/state/execution.go#L210) or failing to [save a block](https://github.com/tendermint/tendermint/blob/0c96f0b434c1d9f23883d1f2291135830a7a5290/store/store.go#L352). In either case we'd panic.

This is also why I've strongly recommended that this interface only be used from full nodes and not validators (we may even want to enforce this if we go with this solution).

The benefit of this approach over just dumping this data into a separate store on disk altogether and having the consumer read from _that_ separate store, however, is that the consumer gets real-time notifications of incoming updates, _and_ they get stronger delivery guarantees than with the subscription interface.

If there's a good database we could use to dump this data into that also provides real-time update notifications to other clients reading from that same database, then that could be a good alternative to this ADR. I've also looked into solutions like using a message broker (e.g. AMQP/Kafka/etc.), but they're all designed to handle small messages (on the order of kilobytes, not megabytes).

Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:224]: Also, with this approach, if the node falls behind the rest of the network, we already have catch-up mechanisms built in to a full node. As the full node catches up, it would stream all of this data to the companion, effectively providing eventual consistency to the companion.

File: docs/architecture/adr-082-data-companion-api.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md]: I've updated the ADR to reflect these answers.

**Comment by JayT106** [comment]
> > Not sure the data scales will be for every CommitBlockRequest. Will it be better if we can filter the events from the Tendermint side?
> 
> I'm not sure I understand @JayT106. Could you please clarify here?

oh, I mean, if the data size of block/tx events or results are enormous, can we discard some data if the companion server doesn't need it? Or just let the companion serve to filter which info it would like to store if we assume the gRPC data transmission won't be an issue?

File: docs/architecture/adr-082-data-companion-api.md (hunk 1 above)
Comment (cmwaters) [reply docs/architecture/adr-082-data-companion-api.md:253]: > Failing to send this data effectively would be the equivalent of failing to save ABCI responses [here](https://github.com/tendermint/tendermint/blob/0c96f0b434c1d9f23883d1f2291135830a7a5290/state/execution.go#L210) or failing to [save a block](https://github.com/tendermint/tendermint/blob/0c96f0b434c1d9f23883d1f2291135830a7a5290/store/store.go#L352). In either case we'd panic.

The main difference, is that this is an external (untrusted) service. The cases you mention are within the domain of the Tendermint node - we know what we're delivering and we can make guarantees around that. I suppose it is possible to say that consensus blocks on saving a block but practically speaking this has different implications than relying on a different process (potentially on a different network).

I agree with William's conceptual framing. The mechanics of channels/buffers mean that if we wan't to fix them in size and don't want to drop data then the process must block for the channel to be emptied. It's one of those triangles where we have to choose 2 of the 3 points. I don't think we should present to users as if we have some silver bullet that does all 3. In this case, the users of this function want to not drop data and want to control how much memory grows, thus they should accept that in unideal circumstances, consensus will block.

I agree with you that failing in a loud way that lets operators know there is a problem instead of simply blocking and waiting makes more sense.

File: docs/architecture/adr-082-data-companion-api.md [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+This data would need to be stored on disk to ensure that Tendermint could
+attempt to resubmit all unsubmitted data upon Tendermint startup. At present,
+this data is already stored on disk, and so practically we would need to
+implement some form of background pruning mechanism to remove the data we know
+has been shared with the companion service.
+
+### Interaction with Block Sync
+
+During block sync, request buffering should be disabled and requests to the
+companion service should be blocking.
```
Comment (cmwaters) [inline docs/architecture/adr-082-data-companion-api.md:264]: I think we should consider have a buffer of 1 just so the block sync protocol is capable of moving on to fetching the next block, while the data service processes the data. I'm assuming the calls are synchronous as the server responds with an error if the operation failed in some way

Comment (cmwaters) [reply docs/architecture/adr-082-data-companion-api.md:264]: Just to be explicit about syncing between the data service and the consensus node. If the data service were somehow to be ahead of the node, it should just skip processing until a request with the next height is sent. If the node is ahead of the data service I'm guessing the data service will error and the node operator would need to roll back the consensus node to the same height (is that right?) or would we persist some amount of responses to disk so we could replay it to the data service?

File: docs/architecture/adr-082-data-companion-api.md [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+  request buffer has been used.
+
+## Implications
+
+1. We will be able to mark the following RPC APIs as deprecated and schedule
+   them for removal in a future release:
+   1. The [WebSocket subscription API][websocket-api]
+   2. [`/tx_search`]
+   3. [`/block_search`]
+   4. [`/broadcast_tx_commit`]
```
Comment (cmwaters) [inline docs/architecture/adr-082-data-companion-api.md:318]: I think this endpoint requires more discussion. I think there's utility in having a tiny index (perhaps builtin to to the blockstore) that is able to let a user know that their transaction has been committed. Perhaps we still deprecate this (as we'd discussed previously) but I think we should still support the use case of I submit a transaction and I poll to check that it is actually committed. Perhaps this requires a broader discussion on transaction tracing

**Review by cmwaters — COMMENTED** [review]
👍 Nice work. I see this progression as an evolution of https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-065-custom-event-indexing.md and satisfies one of the earlier goals of a pluggable indexer (see https://github.com/tendermint/tendermint/issues/7135)

File: docs/architecture/adr-082-data-companion-api.md (hunk 1 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:253]: > The main difference, is that this is an external (untrusted) service.

I'd disagree here. Would you consider a socket-based ABCI app an untrusted external service? Because the configuration for the companion service is specified in the Tendermint config in the same way as for an ABCI socket client.

File: docs/architecture/adr-082-data-companion-api.md [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+self-contained. While this philosophy initially allowed a certain degree of ease
+of node operation (i.e. simple UX), it has also lent itself to feature sprawl,
+with Tendermint being asked to take care of increasingly more than just
+consensus. Under the spotlight in this ADR are:
+
+1. **Event indexing**, which is required in order to facilitate arbitrary block
+   and transaction querying, as well as subscription for arbitrary events. We
+   have already seen, for instance, how the current (somewhat unreliable) event
+   subscription implementation can cause back-pressure into consensus and affect
+   IBC relayer stability (see [\#6729] and [\#7156]).
```
Comment (08d2) [inline docs/architecture/adr-082-data-companion-api.md:31]: Blocks and transactions are on-chain, Merkelized data from which events are derived. Querying block and transaction data doesn't really involve events at all. But this clause (and others) seem to imply some level of equivalence between these two things. Can you maybe speak a bit more on your understanding of events vs. other types of data?

File: docs/architecture/adr-082-data-companion-api.md [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+   of a configuration parameter. When off, it must have negligible (ideally no)
+   impact on system performance.
+
+3. It must not cause back-pressure into consensus.
+
+4. It must not cause unbounded memory growth.
+
+5. Data must be published reliably. If data cannot be published, the node must
+   rather crash in such a way that bringing the node up again will cause it to
+   attempt to republish the unpublished data.
```
Comment (08d2) [inline docs/architecture/adr-082-data-companion-api.md:131]: How do you define "published"? Is it enough that an event is successfully sent onto a Go channel? Written to a socket? Acked at the TCP level?

You can basically pick either reliable delivery _or_ not affecting liveness, but not really both. (Crashing certainly impacts liveness.) And if you pick reliable delivery, you really really want to do that in-process.

File: docs/architecture/adr-082-data-companion-api.md (hunk 5 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:31]: No, I don't see any equivalence between them. I picked those two kinds of data because I know they're important to different kinds of users.

They're also only examples of the kinds of data that we could expose via this interface.

File: docs/architecture/adr-082-data-companion-api.md (hunk 6 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:131]: My assumption here in this document (as I've tried to articulate, but perhaps I didn't do a good job of it) is that users of this interface will be running full nodes (not validators) whose sole purpose is to obtain real-time access to the data this interface exposes.

These consumers would require reliable delivery of the data. That is, if the node cannot deliver the data to the companion, the node must fail.

Think of it almost like a storage device for the node. If the node cannot access its storage, it should crash.

Using gRPC, once the node receives a success response from the companion, the node considers the data to be delivered.

And remember, this interface is opt-in. It's off by default.

File: docs/architecture/adr-082-data-companion-api.md [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+
+This ADR proposes to effectively offload some data storage responsibilities and
+functionality to **single external companion service** in order to:
+
+1. Improve core consensus stability.
+2. Eventually reduce the amount of functionality for which the core team is
+   responsible, thereby improving maintainability.
+3. Still cater for certain use cases that require this data and functionality.
+
+The way the system is currently built is such that a Tendermint node is mostly
```
Comment (ethanfrey) [inline docs/architecture/adr-082-data-companion-api.md:21]: This was fed by a belief that all users would run their own node and use CLI. I fought for thinking web-scale and hosted servers, and using rpc/apis with client libs as first class citizens. This first materialised in the LCD, which was a minimal first step, and I think web-first only really came into being with CosmJS and tools like telescope.

Before making any changes in the direction the ADR suggests, we need to older maintainers, especially @ebuchman , to agree that ease of self-hosting is a secondary goal, and the primary goal is scalability, stability, and features when managed by professional DevOps.

Once there is full agreement on that, we can work on a very clear design for that goal. If that is not supported, and self-hosting is considered important, then that should be clear. Working out that fundamental design decision via various details in this ADR leads to a compromise that does nothing right.

Comment (ethanfrey) [reply docs/architecture/adr-082-data-companion-api.md:21]: If there is clear agreement on a radical change of tendermint and sdk to web-scale, then I would be happy to join some discussion. I know @tac0turtle has done a lot of work and thinking in that direction.

File: docs/architecture/adr-082-data-companion-api.md [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+self-contained. While this philosophy initially allowed a certain degree of ease
+of node operation (i.e. simple UX), it has also lent itself to feature sprawl,
+with Tendermint being asked to take care of increasingly more than just
+consensus. Under the spotlight in this ADR are:
+
+1. **Event indexing**, which is required in order to facilitate arbitrary block
+   and transaction querying, as well as subscription for arbitrary events. We
```
Comment (08d2) [inline docs/architecture/adr-082-data-companion-api.md:28]: Why is it necessary to index events in order to facilitate querying blocks and transactions? Events are not represented in either of those entities.

File: docs/architecture/adr-082-data-companion-api.md [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+   subscription implementation can cause back-pressure into consensus and affect
+   IBC relayer stability (see [\#6729] and [\#7156]).
+2. **Block execution result storage**, which is required to facilitate a number
+   of RPC APIs, but also results in storage of large quantities of data that is
+   not critical to consensus. This data is, however, critical for certain use
+   cases outside of consensus.
+
+Another intersecting issue is that there may be a variety of use cases that
+require different data models. A good example of this is the mere existence of
+the [PostgreSQL indexer] as compared to the default key/value indexer.
```
Comment (08d2) [inline docs/architecture/adr-082-data-companion-api.md:39]: Is it reasonable to expect Tendermint to directly accommodate every use case with sufficient demand?

File: docs/architecture/adr-082-data-companion-api.md [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+      2. Transaction results
+   2. Committed block data
+
+2. It must be opt-in. In other words, it is off by default and turned on by way
+   of a configuration parameter. When off, it must have negligible (ideally no)
+   impact on system performance.
+
+3. It must not cause back-pressure into consensus.
+
+4. It must not cause unbounded memory growth.
```
Comment (08d2) [inline docs/architecture/adr-082-data-companion-api.md:127]: These two requirements are mutually exclusive.

File: docs/architecture/adr-082-data-companion-api.md (hunk 5 above)
Comment (08d2) [reply docs/architecture/adr-082-data-companion-api.md:31]: Purely in terms of information, and ignoring delivery mechanics, would these requirements not be satisfied by polling for the latest block in a hot loop?

Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:31]: Yes, but with the recommended approach in this ADR Tendermint can automatically prune some of this data once it has assurances that the companion's received it. Consider the companion a "privileged" consumer of Tendermint data. Also, consider the possibility of this allowing us to offload responsibility for the majority of the RPC entirely to an external process that would store/index/present data and scale in whatever way users choose, independently of the Tendermint node.

Storage costs are incredibly high for operators at present, and if there's no need to store a particular kind of data to facilitate Tendermint's core functionality, why store it? Especially if someone wants to store/index it in a different way to how Tendermint natively stores/indexes it - in that case, operators would be paying those storage costs twice.

File: docs/architecture/adr-082-data-companion-api.md (hunk 7 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:21]: > Before making any changes in the direction the ADR suggests, we need to older maintainers, especially @ebuchman , to agree that ease of self-hosting is a secondary goal, and the primary goal is scalability, stability, and features when managed by professional DevOps.

Could you please elaborate on how you see self-hosting to be a contrasting goal to management by professional DevOps? Can professional a DevOps team not self-host, for instance? Or are you saying that the original goal was to allow people without any professional DevOps experience to host a node?

> If there is clear agreement on a radical change of tendermint and sdk to web-scale, then I would be happy to join some discussion.

Given that "web-scale" is a somewhat overloaded term and can cause confusion in terms of expectations if we were to make it a goal, I think it would be preferable to focus on solving specific problems. In this case, if users do want to scale their RPC independently of their full node/validator, then this approach gives them the option.

Are there other aspects of the system that you would want to see scaled out independently of the core consensus functionality? We'd be more than happy to discuss synchronously sometime.

File: docs/architecture/adr-082-data-companion-api.md (hunk 8 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:28]: The `/tx`, `/tx_search` and `/block_search` endpoints, for instance, all require the indexers to be able to work.

See https://github.com/tendermint/tendermint/blob/9d01a6880e38b79ec31849e436677c70ccd0e588/rpc/core/tx.go#L20

https://github.com/tendermint/tendermint/blob/9d01a6880e38b79ec31849e436677c70ccd0e588/rpc/core/tx.go#L54

https://github.com/tendermint/tendermint/blob/9d01a6880e38b79ec31849e436677c70ccd0e588/rpc/core/blocks.go#L197

Other endpoints, like getting the block by its associated height, do not require the indexer.

File: docs/architecture/adr-082-data-companion-api.md (hunk 9 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:39]: Not necessarily. If there's sufficient demand for functionality that we consider to be outside of "core" Tendermint functionality, then we should find some way of accommodating that, but that doesn't mean we need to implement and maintain it ourselves.

This is what this ADR attempts to do.

We're a small team with limited capacity. In order for us to meet our mid- to long-term correctness goals for the system, we believe we need to find ways in the short term to trim away and outsource certain parts of the system that we believe are not core to the system's operations - especially parts of the system where Tendermint's effectively catering for application-level concerns.

**Comment by thanethomson** [comment]
I've updated the ADR now based on the feedback so far and have rebased it on `main`.

File: docs/architecture/adr-082-data-companion-api.md (hunk 5 above)
Comment (08d2) [reply docs/architecture/adr-082-data-companion-api.md:37]: > Consider the companion a "privileged" consumer of Tendermint data . . . the approach in this ADR [allows Tendermint to] automatically prune some of this data once it has assurances that the companion's received it.

A stream of events (as defined by this ADR) are derived from and (optimistically) describe on-chain state, but don't, and can't, provide, at a theoretical level, the consistency guarantees made by Tendermint via direct queries of that state, and so can't be treated as equivalent.

If the goal of the ADR is to reduce storage costs of Tendermint nodes, under the assumption that third party consumers will be reading and indexing data on a per-height basis, then a far more effective approach would be to introduce a manual per-height "purge" endpoint. (Maybe it even already exists! I don't know.) "Privileged" consumers of Tendermint data would poll the node for the latest height, fetch that height's block data through query RPCs that, unlike events streams, can be trusted to be authoritative, and then would call that "purge" endpoint when they've indexed and persisted the corresponding data.

File: docs/architecture/adr-082-data-companion-api.md (hunk 8 above)
Comment (08d2) [reply docs/architecture/adr-082-data-companion-api.md:34]: I understood that indexing of transactions is different to indexing of events. Do I misunderstand the context?

File: docs/architecture/adr-082-data-companion-api.md (hunk 7 above)
Comment (08d2) [reply docs/architecture/adr-082-data-companion-api.md:27]: > Could you please elaborate on how you see self-hosting to be a contrasting goal to management by professional DevOps? Can professional a DevOps team not self-host, for instance? Or are you saying that the original goal was to allow people without any professional DevOps experience to host a node?

+1 — "professional DevOps" teams are going to "self-host" nodes by definition, right? 

Maybe the distinction you're trying to draw here, @ethanfrey, is between professional/organizational operators and hobbyist/individual operators? If so, I'm on board. Self-hosting by hobbyist or individual operators is hopefully a secondary, nice-to-have goal of the project; self-hosting by professional or organizational operators is hopefully a primary and mandatory goal of the project.

File: docs/architecture/adr-082-data-companion-api.md (hunk 5 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:37]: > a far more effective approach

Effective to what end? If we want to eventually offload responsibility entirely for providing an RPC interface to a separate process, and Tendermint only maintain the bare minimum set of RPC calls necessary for operational and transaction submission purposes, then the approach you're recommending is not effective.

If we wanted to continue having to maintain our own full RPC as we currently do, then your suggested approach could be effective.

File: docs/architecture/adr-082-data-companion-api.md (hunk 7 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:27]: Afaict, catering to self-hosting by professional/highly skilled/organizational operators of validators is definitely more important than facilitating that for hobbyists, unless someone has a really compelling argument otherwise.

File: docs/architecture/adr-082-data-companion-api.md (hunk 8 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:34]: Sorry, yes, not sure why I said the `/tx` endpoint too. I should've just posted the `/tx_search` and `/block_search` endpoints. Both of those require events to work.

File: docs/architecture/adr-082-data-companion-api.md (hunk 7 above)
Comment (ethanfrey) [reply docs/architecture/adr-082-data-companion-api.md:27]: > Or are you saying that the original goal was to allow people without any professional DevOps experience to host a node?

Yes, the primary focus was on hobbyists running a full node (not a validator) with minimal work. 

This idea seems to be about breaking it into multiple services to allow faster development speed, use of optional services not needed on validator nodes, more features and less attack vectors on (minimal) validator nodes.

I see that as a *good thing* but if we still insist any hobbyist should be able to run a server with solid events indexer, best apis, etc this will block much progress down this road. Unless we just provide eg docker compose for hobbyist that provides a recommended setup including external services (external dbs, messaging queues, api servers, etc)

Basically requiring 3+ services to run and work together, which I think is needed for any decently functional web tech, will quickly hit alarms of "hobbyist can't run it". I just want to highlight this tension early on and seek a resolution of it before more time gets spent on ADRs or worse implementation, which will be blocked on such grounds

**Comment by tac0turtle** [comment]
huge fan of this and something, I would like to help accelerate. For numia, this sort of process is something we have been talking about and are looking to see how fast we can implement.

**Comment by thanethomson** [comment]
> huge fan of this and something, I would like to help accelerate. For numia, this sort of process is something we have been talking about and are looking to see how fast we can implement.

Glad to hear that :slightly_smiling_face: 

> is there a way to define a streamer instead of making it a request response model. Then periodically the consumer can send what heights it has and the node would batch delete them. This can be optional, but reduces complexity of consumers.

Could you please elaborate on the problem that you see with the request/response model?

The way the proposed architecture works is such that the moment that the node gets a success response from the companion for a specific height's data, it can safely delete the non-critical parts of that data (or add it to a batch of data to delete on some periodic schedule).

> I think there is a large refactor of how things are stored needed in order to see the best results

Could you please expand on what you mean here?

File: docs/architecture/adr-082-data-companion-api.md [hunk 11]
```diff
@@ -0,0 +1,425 @@
...
+the [PostgreSQL indexer] as compared to the default key/value indexer.
+
+Continuing from and expanding on the ideas outlined in [RFC 006][rfc-006], the
+suggested approach in this ADR is to provide a mechanism to publish certain data
+from Tendermint, in real-time and with certain reliability guarantees, to a
+single companion service outside of the node that can use this data in whatever
+way it chooses (filter and republish it, store it, manipulate or enrich it,
+etc.).
+
+Specifically, this mechanism would initially publish:
```
Comment (cmwaters) [inline docs/architecture/adr-082-data-companion-api.md:54]: Just in regards to the `initially` part: There isn't any other (global) data that I know of that Tendermint could possibly publish - this is basically everything

**Review by cmwaters — APPROVED** [review]
💯 Great write up.

File: docs/architecture/adr-082-data-companion-api.md (hunk 7 above)
Comment (williambanfield) [reply docs/architecture/adr-082-data-companion-api.md:27]: I think that this is a pretty fundamental tension within Tendermint that I'd be interested in seeing resolved. Resolving that seems like a rather large task and I'm not certain how to begin it. I would say that I am firmly on the side that hobbyists should not be considered the primary audience for running Tendermint. This tension comes up fairly frequently. 

For example, it comes up when discussing logging within Tendermint. Operators sometimes complain that the logging has a tendency to 'quickly fill their terminal' so tools to filter logging should be built into Tendermint. While I'm sympathetic to not simply spamming the logs with useless nonsense, it does appear to me that common devops practices like running an external log indexer are not necessarily common for operators. Similar tension arises for me when I discuss how to access prometheus metrics with operators who don't always seem to be running a prom instance. If we want to make this shift in direction which, as @ethanfrey deftly points out, will allow faster development, we may also have to do a decent amount of documentation / explanation / evangelism to clarify why this is preferred.

File: docs/architecture/adr-082-data-companion-api.md [hunk 12]
```diff
@@ -0,0 +1,425 @@
...
+   IBC relayer stability (see [\#6729] and [\#7156]).
+2. **Block execution result storage**, which is required to facilitate a number
+   of RPC APIs, but also results in storage of large quantities of data that is
+   not critical to consensus. This data is, however, critical for certain use
+   cases outside of consensus.
+
+Another intersecting issue is that there may be a variety of use cases that
+require different data models. A good example of this is the mere existence of
+the [PostgreSQL indexer] as compared to the default key/value indexer.
+
```
Comment (lasarojc) [inline docs/architecture/adr-082-data-companion-api.md:46]: I strongly support offloading the indexing because, imho, indexing should only concern the application.
In this sense, I fear that the proposed approach is still too tied to Tendermint. One possibility to resolve this would be to place the companion in the application level, i.e., above ABCI, and provide an ABCI shim that intercepts and indexes the events.

```
      +----------+
      | ABCI App |
      +----------+ 
             Î
      +----------+.         +-----------------+
      | ABCI Shim |. ---> | Data Companion |
      +----------+          +-----------------+
             Î           
     +------------+
     | Tendermint | 
     +------------+
 ```

The same shim/interceptor could be used later to provide a default app-side mempool, for example.

Block execution will be slowed down by the indexing, but the coupling is inevitable if we want to tie the success of the index operation with the progress of a tendermint node.

File: docs/architecture/adr-082-data-companion-api.md [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+      single consumer and single event window.
+
+   Making the requisite modifications to ADR-075 basically converges on the
+   solution outlined in this ADR, except that ADR-075 publishes data via JSON over
+   HTTP while this solution proposes gRPC.
+
+2. We could pick a database that provides real-time notifications of inserts to
+   consumers and just dump the requisite data into that database. Consumers
+   would then need to connect to that database, listen for updates, and then
+   transform/process the data as needed.
```
Comment (lasarojc) [inline docs/architecture/adr-082-data-companion-api.md:94]: Events are essentially a stream of never repeating data and in which pruning is efficient and compaction is pretty much useless. We should investigate using very simple stream DBs; the ideal one wouldn't be very different from a WAL.

File: docs/architecture/adr-082-data-companion-api.md [hunk 14]
```diff
@@ -0,0 +1,425 @@
...
+   This approach could also inevitably lead to feature sprawl in people wanting
+   us to support multiple different databases. It is also not clear which
+   database would be the best option here, as that would most likely be
+   influenced by integrators' use cases.
+
+   Beyond this, if we exclusively support one database but integrators needed
+   another for their use case, they would be forced to perform some kind of
+   continuous ETL operation to transfer the data from the one database to the
+   other, potentially doubling their storage capacity requirements and therefore
+   their storage-related costs. This would also increase pruning complexity.
```
Comment (lasarojc) [inline docs/architecture/adr-082-data-companion-api.md:105]: the ABCI interceptor approach would be useful here, since app developers could just drop the interceptor and handle the indexing at the app, according to their needs.

File: docs/architecture/adr-082-data-companion-api.md [hunk 15]
```diff
@@ -0,0 +1,425 @@
...
+# Controls the BlockCommitted gRPC call.
+[data_companion.block_committed]
+# Enable the BlockCommitted gRPC call. Only relevant if the data companion is
+# enabled.
+enabled = true
+# Additional fields to publish in each BlockCommittedRequest sent to the
+# companion. Available options:
+# - "finalize_block_response": Also publish the FinalizeBlock response related
+#                              to the block in the BlockCommittedRequest.
+additionally_publish = ["finalize_block_response"]
```
Comment (lasarojc) [inline docs/architecture/adr-082-data-companion-api.md:308]: Why not a single list of data to be published, instead of `data_companion.block_committed` and `additionally_publish`?

File: docs/architecture/adr-082-data-companion-api.md (hunk 12 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:46]: Very interesting idea!

Right now, the design of this ADR leverages the fact that Tendermint currently persists event data to disk, so it can reliably push that data to the companion (i.e. on-disk buffering). Is your idea that the ABCI shim would instead offer its own on-disk buffering?

Also, the proposed architecture in ADR-082 allows for sending of full block data as well. Is your idea to have the shim and the Tendermint node each make a connection through to the companion to be able to pass these different types of data through?

File: docs/architecture/adr-082-data-companion-api.md (hunk 13 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:94]: So basically a time-series database, like InfluxDB?

File: docs/architecture/adr-082-data-companion-api.md (hunk 15 above)
Comment (thanethomson) [reply docs/architecture/adr-082-data-companion-api.md:308]: You're right, that's probably easier to deal with from a UX perspective. Instead of enabling different RPC calls (an implementation detail), rather let the user specify which kinds of data they want and let the companion client figure out which corresponding RPC calls and data structures to send.

I'll change the interface here.

File: docs/architecture/adr-082-data-companion-api.md (hunk 12 above)
Comment (lasarojc) [reply docs/architecture/adr-082-data-companion-api.md]: Yes, whatever is needed to implement the communication with the data companion, would be done at the Shim/Companion level. And any data flowing through ABCI, be it blocks going up or events going down, could be intercepted and sent to the companion.
